nifi-nar-bundles/nifi-adx-bundle/nifi-adx-processors/src/main/java/org/apache/nifi/processors/adx/AzureAdxSourceProcessor.java

```diff
@@ -0,0 +1,192 @@
+        executionClient = service.getKustoQueryClient();
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile outgoingFlowFile;
+        String databaseName = context.getProperty(DB_NAME).getValue();
+        String adxQuery;
+        KustoOperationResult kustoOperationResult;
+
+        //checks if this processor has any preceding connection, if yes retrieve
+        if ( context.hasIncomingConnection() ) {
+            FlowFile incomingFlowFile = session.get();
+            //incoming connection exists but the incoming flowfile is null
+            if ( incomingFlowFile == null && context.hasNonLoopConnection() ) {
+                getLogger().error("Incoming connection exists but Incoming flow file is null");
+                return;
+            }
+            //incoming connection exists and retrieve adxQuery from context
+            if ( incomingFlowFile != null && incomingFlowFile.getSize() == 0 ) {
+                if ( context.getProperty(ADX_QUERY).isSet() ) {
+                    adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = session.putAttribute(incomingFlowFile, ADX_QUERY_ERROR_MESSAGE, message);
+                    session.transfer(incomingFlowFile, RL_FAILED);
+                    return;
+                }
+            } else {
+                try {
+                    adxQuery = getQuery(session, incomingFlowFile);
+                } catch(IOException ioe) {
+                    getLogger().error("Exception while reading from FlowFile " + ioe.getLocalizedMessage(), ioe);
+                    throw new ProcessException(ioe);
+                }
+            }
+            outgoingFlowFile = incomingFlowFile;
+        } else {
+            outgoingFlowFile = session.create();
+            adxQuery = context.getProperty(ADX_QUERY).evaluateAttributeExpressions(outgoingFlowFile).getValue();
+        }
+
+        try {
+            //execute Query
+            kustoOperationResult = executeQuery(databaseName,adxQuery);
+            List<List<Object>> tableData = kustoOperationResult.getPrimaryResults().getData();
+            String json = tableData.size() == 1 ? objectMapper.writeValueAsString(tableData.get(0)) : objectMapper.writeValueAsString(tableData);
+            if ( getLogger().isDebugEnabled() ) {
+                getLogger().debug("Query result {} ", tableData);
+            }
+            try(ByteArrayInputStream bais = new ByteArrayInputStream(json.getBytes())){
+                session.importFrom(bais, outgoingFlowFile);
+            }
```

Comment:
This type of approach should be changed to use stream the data to an OutputStream using ObjectMapper, instead of holding it all in memory.